changefeedtrigger/__init__.py

```diff
@@ -0,0 +1,94 @@
+import time
+import json
+from subprocess import Popen, PIPE
+from azure.storage.blob.changefeed import ChangeFeedClient
+from azure.storage.blob import BlobServiceClient
+import azure.functions as func
+from datetime import timedelta, datetime
+import logging
+from predicate import where
+
+
+def multidimensional_filtering(cf_client, b_filter, blob_client, cont_name, c_filter, e_filter):
+    filter_event_service = blob_client.get_container_client(container=cont_name)
+    blob_name = "continuation-token"
+
+    # access blobs in container
+    blob_list = filter_event_service.list_blobs()
+    tokens = []
+    for blob in blob_list:
+        tokens.append(blob.name)
+    
+    # check if continuation token already exists
+    if blob_name in tokens:
+        cursor = filter_event_service.download_blob(blob_name).readall()
+        # if value of token is None return None
+        if cursor == b"None":
+            return None
+        # if value of token is a cursor initiate the changefeed with cursor
+        elif cursor != b"" and cursor != b"None":
+            change_feed = cf_client.list_changes(results_per_page=500).by_page(continuation_token=eval(cursor))
+        # if token is empty initiate changefeed without cursor
+        else:
+            change_feed = cf_client.list_changes(results_per_page=500).by_page()
+    # if token does not exist upload token as empty string placeholder as blob and initiate changefeed without cursor
+    else:
+        filter_event_service.upload_blob(blob_name, "") 
+        change_feed = cf_client.list_changes(results_per_page=500).by_page()
+
+    # iterate through change feed
+    change_feed_page = next(change_feed)
+
+    # get multidimensionally filtered events
+    #filtered_events = list(filter(where(data__eventType__iexact=e_filter, data__subject__icontains=("/" + c_filter + "/blobs/" + b_filter)), events))
+    filtered_events = []
+    for event in change_feed_page:
+        if (event["eventType"] == e_filter) and (("/" + c_filter + "/") in event["subject"]) and (("/" + b_filter) in event["subject"]):
```

Comment:
this should be passed as a predicate into function.

Reply:
i was running into an issue previously with using a predicate that i could only find a method for "if contains/icontains" for integers and not strings, so although i was able to use a predicate to filter the eventType using a "if exact/iexact", i was unable to filter blob or container name. i figured that it would have been more beneficial for the client for the program to be able to filter those items rather than to use a predicate, but let me know what you think!

Reply:
I've written a snippet that illustrates what's in my mind. Let me know if this helps:
```


event1 = {
    "name" : "aaa",
    "id" : 1,
    "type" : "x"
}

event2 = {
    "name" : "bbb",
    "id" : 2,
    "type" : "y"
}

event3 = {
    "name" : "bbb",
    "id" : 3,
    "type" : "z"
}

events = [event1, event2, event3]

print(events)

def is_event_2(evt):
    return evt["id"] == 2
    
filtered_events = filter(is_event_2, events)

print(list(filtered_events))

filtered_events2 = filter(lambda e: e["name"] == "bbb" and e["type"] == "z", events)

print(list(filtered_events2))
```

Reply:
I think the issue is that when the changefeed iterates on something it looks like this:

{'schemaVersion': 4, 'topic': '/subscriptions/ba45b233-e2ef-4169-8808-49eb0d8eba0d/resourceGroups/carmen_romero/providers/Microsoft.Storage/storageAccounts/changefeedtester', **'subject': '/blobServices/default/containers/test-changefeed-container/blobs/blob2**', **'eventType': 'BlobCreated',** 'eventTime': '2020-08-07T18:40:47.5471927Z', 'id': 'ab733b55-b01e-005d-5dea-6c750f0649b7', 'data': {'api': 'PutBlob', 'clientRequestId': '81d5eb64-d8dd-11ea-acd5-3c58c280e73c', 'requestId': 'ab733b55-b01e-005d-5dea-6c750f000000', 'etag': '0x8D83B01669CA037', 'contentType': 'application/octet-stream', 'contentLength': 11, 'blobType': 'BlockBlob', 'blobVersion': None, 'containerVersion': '01D66C4E130F7BB1', 'blobTier': None, 'url': '', 'sequencer': '00000000000000000000000000001DBF00000000006cae79', 'previousInfo': None, 'snapshot': None, 'blobPropertiesUpdated': None, 'asyncOperationInfo': None, 'storageDiagnostics': {'bid': 'a3cc2f19-a006-000c-00ea-6ce883000000', 'sid': '4a4bcf2e-d709-ebc1-44a7-1af56aff1320', 'seq': '(7615,1401396,7122553,6626588)'}}}

I bolded the "eventType" and the "subject", which is where we can locate the blob name and container name. So the issue is that we can't directly access the blob name and container like we can for the event type, we have to enter "subject" and check to see if the string "exampleBlobName" is anywhere in "subject"-- same thing for the container.

So, I think maybe the solution you put would be able to do this if we could do .contains instead of == in matches(), but i am not sure. Let me know what you think.

Reply:
You should be able to substitute == with whatever is suitable. The predicate is just function returning boolean so anything that evaluates to boolean is fine (like anything put in if() statement can be converted into predicate).

Reply:
okay that makes sense! i will start working on implementing this